scripts/16-rust-testing.md

```diff
@@ -0,0 +1,647 @@
+        WebEvent::PageUnload => println!("page unloaded"),
+        // Destructure `c` from inside the `enum`.
+        WebEvent::KeyPress(c) => println!("pressed {c}."),
+        // Destructure `Click` into `x` and `y`.
+        WebEvent::Click { x, y } => {
+            println!("clicked at x={x}, y={y}.");
+        },
+    }
+}
+```
+
+notes:
+
+Then in our code, instead of writing spaghetti if statements, we match the current state of the application, and execute different behaviours based on this state.
+No other actions are possible, because we're responding to exactly the state we are given.
+If more nuance is needed, then don't write an if statement, add more detail to the model.
+
+This keeps us safe.
+
+You might not have noticed that we've not handled all cases of the WebEvent enum.
+Paste is unhanded.
+It's easy for a human to miss, even in this little example.
+
+But the compiler didn't miss it.
+
+---
+
+```js
+error[E0004]: non-exhaustive patterns:
+  --> src/main.rs:24:11  
+|
+|    match event {  
+|          ^^^^^ pattern `WebEvent::Paste(_)` not covered
+  
+help: ensure that all possible cases are being handled by
+adding a match arm with a wildcard pattern or
+an explicit pattern as shown
+| 
+|        WebEvent::Paste(_) => todo!(),  
+|
+```
+
+notes:
+
+If you model your whole application state using enums and structs it becomes very difficult to make logic errors.
+The incredibly thorough compiler, powered by the extra syntax Rust has compared to other languages, allows us to express our intentions in a machine-readable way.
+
+I explained this in more detail in my previous video "Rust Makes Cents", so for now, we will move on from language to tooling.
+
+---
+
+# Clippy
+
+```sh
+cargo clippy --fix -- \
+-W clippy::pedantic \
+-W clippy::nursery \
+-W clippy::unwrap_used \
+```
+
+notes:
+
+When used right, rust's built-in linter won't just make your code cleaner and more idiomatic, but with the unwrap_used warning, here, safer and more correct, too.
+
+The unwrap_used warning reminds you that while `.unwrap()`ing a result is fine for prototype code, you must not let it creep into your production code.
+
+In fact, I recommend in your CI pipeline, you configure this warning to be an error, and fail the build.
+Force you and your team to explain why they are so sure the result is safe, using the `.expect("reason")` method.
+
+More details in my previous video about Rust errors.
+
+---
+
+```sql
+$ cargo clippy --fix
+```
+
+```md
+error:
+the working directory has uncommitted changes,
+and `cargo fix` can potentially perform destructive changes
+if you'd like to suppress this error pass `--allow-dirty`,
+`--allow-staged`, or commit the changes to these files
+```
+
+notes:
+
+If you run clippy with the --fix option, which can change your code if it is safe to do so, by default you will see this warning if you are not checked-in in version control.
+
+The cargo developers really have thought of everything!
+
```

Comment:
Given the amount of content in this video, this section is somewhat repetitive of previous videos and should be summarised quickly.

My take away of this section, in the context of testing is more about reducing the need for testing, so that you can focus your testing on the areas that matter most.

This is good.

I'm not an advocate of performing blanket testing, or hunting an illusive 100% coverage report. So easy wins like a linter enforcing `#![forbid(unsafe_code)]` and focusing our efforts on the high-impact testing or refactors, such as what the [Safety Dance](https://github.com/rust-secure-code/safety-dance/) team do is good.

Reply:
As with all youtubers, I somewhat have to assume viewers may have missed any and all of my previous videos.

Reply:
Understood! It's still a good section, it just seems to get a disproportionate amount of screen time. I see you are looking at adding some additional content later, so that will fix this.